DPC++ compile bug

In file `portBLAS/include/blas_meta.h`, the header file 
```cpp
#include <ext/oneapi/experimental/sycl_complex.hpp>
```
no longer exists in DPC++, instead you can use 
```cpp
#include <ext/oneapi/experimental/complex/complex.hpp>
```
which is located in `llvm-dpcpp/build/include/sycl/ext/oneapi/experimental/complex/complex.hpp`.

I also find another compile error when I use DPC++ on NVIDIA A100, but I can't fix it.

I want to test the performance of portBLAS gemm on tensor core, so I revise `cmake/Modules/FindDPCPP.cmake` and add 2 lines to set the macro NVIDIA_GPU
```cmake
  if (${start_idx} AND ${sm_val} GREATER_EQUAL "80")
    add_definitions(-DSB_ENABLE_JOINT_MATRIX=1)
    add_definitions(-DNVIDIA_GPU=1)
    list(APPEND DPCPP_FLAGS "-Xclang;-cl-mad-enable")
    list(APPEND DPCPP_FLAGS "-DSYCL_EXT_ONEAPI_MATRIX_VERSION=4")
    list(APPEND DPCPP_FLAGS "-DSB_ENABLE_JOINT_MATRIX=1")
    list(APPEND DPCPP_FLAGS "-DNVIDIA_GPU=1")
  endif()
```
Meanwhile, I revise the call of API `joint_matrix_mad` in `src/operations/blas3/gemm_local_joint_matrix.hpp:830` according to the update of DPC++
```cpp
        // reg_res[frag] = joint_matrix_mad(sg, inA, inB, reg_res[frag]);
        joint_matrix_mad(sg, reg_res[frag], inA, inB, reg_res[frag]);
```
After that I compile portBLAS and sample with the command 
```shell
$ CC=clang CXX=clang++ cmake -GNinja ../ -DSYCL_COMPILER=dpcpp -DDPCPP_SYCL_TARGET="nvptx64-nvidia-cuda" -DDPCPP_SYCL_ARCH="sm_80" -DCMAKE_PREFIX_PATH=/opt/OpenBLAS -DCMAKE_THREAD_LIBS_INIT=-lpthread -DBLAS_ENABLE_TESTING=OFF -DBLAS_ENABLE_BENCHMARK=OFF -DCMAKE_BUILD_TYPE=Debug
$ ninja
```
and get the error
```shell
portBLAS/samples/../src/operations/blas3/gemm_local_joint_matrix.hpp:562:13: error: use of undeclared identifier 'get_wi_data'
  562 |             get_wi_data(sg, float_out)[i] = alpha_ * data_left;
      |             ^
portBLAS/samples/../src/operations/blas3/gemm_local_joint_matrix.hpp:607:9: error: use of undeclared identifier 'get_wi_data'
  607 |         get_wi_data(sg, float_out)[i] =
      |         ^
portBLAS/samples/../src/operations/blas3/gemm_local_joint_matrix.hpp:576:40: error: use of undeclared identifier 'get_wi_data'
  576 |                 static_cast<element_t>(get_wi_data(sg, reg_res[frag])[i]);
      |                                        ^
```

In DPC++ `sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp`, the call of `sycl::ext::oneapi::detail::get_wi_data` is used for Intel GPU, while for NVIDIA GPU and AMD GPU, sycl visits the object in joint_matrix jm by directly call `jm.matrix_impl.wi_marray`, but it doesn't work for portBLAS. If I define the macro
```cpp
#define get_wi_data(sg, jm) jm.matrix_impl.wi_marray
```
and compile, I will get another error
```shell
portBLAS/samples/../src/operations/blas3/gemm_local_joint_matrix.hpp:562:13: error: no member named 'matrix_impl' in 'sycl::ext::oneapi::experimental::matrix::joint_matrix<sycl::sub_group, float, sycl::ext::oneapi::experimental::matrix::use::accumulator, 16, 16>'
  562 |             get_wi_data(sg, float_out)[i] = alpha_ * data_left;
      |             ^               ~~~~~~~~~
portBLAS/samples/../src/operations/blas3/gemm_local_joint_matrix.hpp:33:32: note: expanded from macro 'get_wi_data'
   33 | #define get_wi_data(sg, jm) jm.matrix_impl.wi_marray
      |                             ~~ ^
```

I want to learn how to fix it and test the performance of gemm on tensor core.

--- a/.github/workflows/build-and-test.yml
+++ b/.github/workflows/build-and-test.yml
@@ -16,7 +16,7 @@ jobs:
 # cancel outdated builds on pull requests.
   skip-check:
     continue-on-error: true
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
     # Map a step output to a job output
     outputs:
       should_skip: ${{ steps.skip_check.outputs.should_skip }}
@@ -34,7 +34,7 @@ jobs:
   Build-and-Test:
     needs: skip-check
     if: ${{ needs.pre_job.outputs.should_skip != 'true' }}
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
     strategy:
       fail-fast: false
       matrix:

--- a/.scripts/build_dpcpp.sh
+++ b/.scripts/build_dpcpp.sh
@@ -5,7 +5,7 @@ set -ev
 ###########################
 # Get DPCPP
 ###########################
-wget --no-verbose https://github.com/intel/llvm/releases/download/sycl-nightly/20230727/dpcpp-compiler.tar.gz -O dpcpp-compiler.tar.gz
+wget --no-verbose https://github.com/intel/llvm/releases/download/nightly-2023-12-06/sycl_linux.tar.gz -O dpcpp-compiler.tar.gz
 rm -rf /tmp/dpcpp && mkdir /tmp/dpcpp/
 tar -xzf dpcpp-compiler.tar.gz -C /tmp/dpcpp --strip-components 1
 ls -R /tmp/dpcpp/

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,4 +1,4 @@
-FROM ubuntu:focal
+FROM ubuntu:jammy
 
 # Default values for the build
 ARG command

--- a/cmake/CmakeFunctionHelper.cmake
+++ b/cmake/CmakeFunctionHelper.cmake
@@ -616,19 +616,24 @@ else() # default cpu backend
         64 8 8 8 8 1 1 1 1 1 1 1 1 1 float float "no_local" "naive" "none" 1 "strided" "false" "false")
     else()
       add_gemm_configuration(
-        "${data}"  64 "false" "false" "false"
-        64 2 2 8 8 1 1 1 1 1 1 1 1 1 float float "no_local" "standard" "full" 2 "strided" "false" "false")
+        "${data}"  128 "false" "false" "false"
+        64 2 2 2 2 1 1 1 1 1 1 1 1 1 float float "no_local" "standard" "full" 2 "strided" "false" "false")
       add_gemm_configuration(
-        "${data}"  64 "false" "false" "false"
-        64 8 8 8 8 1 1 1 1 1 1 1 1 1 float float "no_local" "standard" "partial" 1 "strided" "false" "false")
+        "${data}"  128 "false" "false" "false"
+        64 4 4 8 8 1 1 1 1 1 1 1 1 1 float float "no_local" "standard" "full" 1 "strided" "false" "false")
+      add_gemm_configuration(
+        "${data}"  128 "false" "false" "false"
+        64 4 4 4 4 1 1 1 1 1 1 1 1 1 float float "no_local" "standard" "partial" 1 "strided" "false" "false")
       add_gemm_configuration(
         "${data}"  64 "false" "false" "false"
         64 2 2 8 8 1 1 1 1 1 1 1 1 1 float float "local" "standard" "full" 2 "strided" "false" "false")
+
     endif()
 
     add_gemm_configuration(
       "${data}" 64 "false" "false" "false"
       64 2 2 4 4 1 1 1 1 4 4 1 1 1 float float "no_local" "standard" "full" 4 "interleaved" "false" "false")
+
   endforeach()
   if(BLAS_ENABLE_COMPLEX)
     # Extract list of complex<data> for each data in supported_types

--- a/src/interface/blas3/backend/default_cpu.hpp
+++ b/src/interface/blas3/backend/default_cpu.hpp
@@ -69,10 +69,10 @@ _gemm(sb_handle_t& sb_handle, index_t _M, index_t _N, index_t _K,
                               _b, _ldb, _strideb, _beta, _c, _ldc, _stridec,
                               batch_size, _dependencies);
 #else
-    if (_M <= 128 && _N <= 128 && _K <= 128 && !s_a && !s_b) {
+    if (_M <= 128 && _N <= 128 && _K <= 256 && !s_a && !s_b) {
       return blas::Gemm_Launcher<
-          container_0_t, container_1_t, container_2_t, 64, false, false, false,
-          64, Tile<2, 2, 8, 8>, _t_a, _t_b, s_a, s_b,
+          container_0_t, container_1_t, container_2_t, 128, false, false, false,
+          64, Tile<2, 2, 2, 2>, _t_a, _t_b, s_a, s_b,
           static_cast<int>(gemm_memory_t::no_local),
           static_cast<int>(gemm_algorithm_t::standard),
           static_cast<int>(gemm_vectorization_t::full), is_beta_zero, 2,
@@ -80,13 +80,24 @@ _gemm(sb_handle_t& sb_handle, index_t _M, index_t _N, index_t _K,
           template _select_gemm(sb_handle, _M, _N, _K, _alpha, _a, _lda,
                                 _stridea, _b, _ldb, _strideb, _beta, _c, _ldc,
                                 _stridec, batch_size, _dependencies);
-    } else if (!s_a && !s_b) {
+    } else if ((_M * _N) >= 524288 && !s_a && !s_b) {
       return blas::Gemm_Launcher<
-          container_0_t, container_1_t, container_2_t, 64, false, false, false,
-          64, Tile<8, 8, 8, 8>, _t_a, _t_b, s_a, s_b,
+          container_0_t, container_1_t, container_2_t, 128, false, false, false,
+          64, Tile<4, 4, 4, 4>, _t_a, _t_b, s_a, s_b,
           static_cast<int>(gemm_memory_t::no_local),
           static_cast<int>(gemm_algorithm_t::standard),
           static_cast<int>(gemm_vectorization_t::partial), is_beta_zero, 1,
+          static_cast<int>(gemm_batch_type_t::strided)>::
+          template _select_gemm(sb_handle, _M, _N, _K, _alpha, _a, _lda,
+                                _stridea, _b, _ldb, _strideb, _beta, _c, _ldc,
+                                _stridec, batch_size, _dependencies);
+    } else if (!s_a && !s_b) {
+      return blas::Gemm_Launcher<
+          container_0_t, container_1_t, container_2_t, 128, false, false, false,
+          64, Tile<4, 4, 8, 8>, _t_a, _t_b, s_a, s_b,
+          static_cast<int>(gemm_memory_t::no_local),
+          static_cast<int>(gemm_algorithm_t::standard),
+          static_cast<int>(gemm_vectorization_t::full), is_beta_zero, 1,
           static_cast<int>(gemm_batch_type_t::strided)>::
           template _select_gemm(sb_handle, _M, _N, _K, _alpha, _a, _lda,
                                 _stridea, _b, _ldb, _strideb, _beta, _c, _ldc,

